core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -3167,6 +3170,687 @@ public void testRegisterExistingTable() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot baseSnapshot = table.currentSnapshot();
+
+    // Create branch and tag from same snapshot
+    String branchName = "etl-pipeline";
+    String tagName = "data-migration-checkpoint";
+
+    table
+        .manageSnapshots()
+        .createBranch(branchName, baseSnapshot.snapshotId())
+        .createTag(tagName, baseSnapshot.snapshotId())
+        .commit();
+
+    // Add data to main branch
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot mainSnapshot2 = table.currentSnapshot();
+
+    // Add data to feature branch
+    table.newFastAppend().toBranch(branchName).appendFile(FILE_C).commit();
+
+    // Verify isolation
+    Map<String, SnapshotRef> refs = table.refs();
+    SnapshotRef mainRef = refs.get(SnapshotRef.MAIN_BRANCH);
+    SnapshotRef branchRef = refs.get(branchName);
+    SnapshotRef tagRef = refs.get(tagName);
+
+    // Main branch should point to latest main snapshot
+    assertThat(mainRef.snapshotId()).isEqualTo(mainSnapshot2.snapshotId());
+
+    // Feature branch should point to its own snapshot (different from main)
+    assertThat(branchRef.snapshotId()).isNotEqualTo(mainRef.snapshotId());
+
+    // Tag should still point to original snapshot
+    assertThat(tagRef.snapshotId()).isEqualTo(baseSnapshot.snapshotId());
+  }
+
+  @Test
+  public void testCreateBranchFailsWhenBranchExists() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+
+    String branchName = "production-pipeline";
+    table.manageSnapshots().createBranch(branchName).commit();
+
+    // Attempt to create the same branch again should fail
+    assertThatThrownBy(() -> table.manageSnapshots().createBranch(branchName).commit())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Ref %s already exists", branchName);
+  }
+
+  @Test
+  public void testCreateTagFailsWhenTagExists() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot currentSnapshot = table.currentSnapshot();
+
+    String tagName = "production-checkpoint";
+    table.manageSnapshots().createTag(tagName, currentSnapshot.snapshotId()).commit();
+
+    // Attempt to create the same tag again should fail
+    assertThatThrownBy(
+            () -> table.manageSnapshots().createTag(tagName, currentSnapshot.snapshotId()).commit())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Ref %s already exists", tagName);
+  }
+
+  @Test
+  public void testRemoveBranchFailsWhenBranchDoesNotExist() {
+    C catalog = catalog();
```

Comment:
These tests aren't hitting the catalog, they're just asserting failures that happen on the client.